packages/python/accern_data/accern_data.py

```diff
@@ -574,74 +627,119 @@ def download_range(
-                start_date,
-                output_path,
-                output_pattern,
-                is_first_time=True,
-                mode=valid_mode,
-                filters=valid_filters,
-                progress_bar=progress_bar)
-        else:
-            is_first_time = True
-            total = 0
-            progress_bar = ProgressBar(
-                total=len(pd.date_range(start_date, end_date)),
-                desc="Fetching info",
-                verbose=verbose)
-
-            for cur_date in pd.date_range(start_date, end_date):
-                expected_records.append(
-                    self._read_total(cur_date, valid_filters))
-                progress_bar.update(1)
-
-            total = sum(expected_records)
-            progress_bar.set_total(total=total)
-            progress_bar.set_description(desc="Downloading signals")
-
-            for ix, cur_date in enumerate(pd.date_range(start_date, end_date)):
-                print_fn(f"now processing {cur_date}")
-                print_fn(f"expected {expected_records[ix]}")
-                progress_bar.set_description(
-                    f"Downloading signals for {cur_date.strftime('%Y-%m-%d')}")
-                is_first_time = self._process_date(
-                    cur_date.strftime("%Y-%m-%d"),
+
+        valid_mode: Optional[Mode] = None
+        cur_date: Optional[pd.Timestamp] = None
+        prev_date: Optional[pd.Timestamp] = None
+        indicator_obj: Optional[ProgressIndicator] = None
+
+        def set_active_mode(
+                mode: Mode,
+                date: pd.Timestamp,
+                indicator: ProgressIndicator) -> None:
+            nonlocal valid_mode, cur_date, indicator_obj
+            valid_mode = mode
+            cur_date = date
+            indicator_obj = indicator
+
+        is_first_time = True
+        for res in self.iterate_range(
+                start_date=start_date,
+                end_date=end_date,
+                mode=mode,
+                filters=filters,
+                chunk_size=None,
+                indicator=indicator,
+                set_active_mode=set_active_mode):
+            assert valid_mode is not None
+            assert cur_date is not None
+            assert indicator_obj is not None
+
+            if cur_date != prev_date:
+                if prev_date is not None:
+                    valid_mode.finish_day(indicator_obj)
+                valid_mode.init_day(
+                    cur_date.strftime('%Y-%m-%d'),
                     output_path,
                     output_pattern,
-                    is_first_time=is_first_time,
-                    mode=valid_mode,
-                    filters=valid_filters,
-                    progress_bar=progress_bar)
-        progress_bar.close()
+                    is_first_time,
+                    indicator_obj)
+            valid_mode.add_result(res)
+            prev_date = cur_date
+            is_first_time = False
```

Comment:
again first time shouldn't be necessary and it should be handled inside the mode. the mode needs to keep track of whether it has already written the header to the file. if you switch files you just reset the cols attribute so you need to wait for a write before it is available again. also, don't return/yield empty dataframes